File: app/src/main/java/com/demo/assignment/core/BaseActivity.java
package com.demo.assignment.core;


import android.os.Bundle;

import androidx.annotation.LayoutRes;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;

public abstract class BaseActivity<B extends ViewDataBinding, VM extends BaseViewModel> extends AppCompatActivity {

    protected VM viewModel;
    protected B binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, getLayoutResId());
        viewModel = createViewModel();
    }

    protected abstract VM createViewModel();

    protected abstract
    @LayoutRes
    int getLayoutResId();

}

File: app/src/main/java/com/digital/playground/core/BaseActivity.kt
package com.digital.playground.core

import android.os.Bundle
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding

/**
 * @Details BaseActivity is abstract class which having the
 * the common properties of all activity.
 * @Author Roshan Bhagat
 */
abstract class BaseActivity<B : ViewDataBinding, VM : BaseViewModel> : AppCompatActivity() {
    protected lateinit var viewModel: VM
    protected lateinit var binding: B

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, layoutResId)
        viewModel = createViewModel()
    }

    protected abstract fun createViewModel(): VM

    @get:LayoutRes
    protected abstract val layoutResId: Int


}